File: python/src/bursar/credits/types/catalog.py
"""Catalog types — mirrors JS SDK's ``credits/types/catalog.ts``."""

from __future__ import annotations

from decimal import Decimal
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field


class CatalogRevision(BaseModel):
    id: str
    config: dict[str, Any]
    version: int


class CatalogRevisionSummary(BaseModel):
    id: str
    version: int
    label: str | None
    active: bool
    created_at: str


class PlanAllowancePolicy(BaseModel):
    model_config = ConfigDict(extra="forbid")

    amount: Decimal
    priority: int = Field(ge=0)
    reset_unit: Literal["second", "minute", "hour", "day", "week", "month", "year"]
    reset_count: int = Field(gt=0)
    reset_anchor: Literal["calendar", "plan_assignment", "rolling"]
    reset_timezone: str


class PlanCreditPolicy(BaseModel):
    model_config = ConfigDict(extra="forbid")

    type: Literal["prepaid", "credit_line"]
    credit_limit: Decimal | None


class _PlanAdmissionOperation(BaseModel):
    model_config = ConfigDict(extra="forbid")

    max_in_flight: int | None = Field(default=None, gt=0)


class PlanAdmissionPolicy(BaseModel):
    model_config = ConfigDict(extra="forbid")

    max_in_flight: int | None = Field(default=None, gt=0)
    operations: dict[str, _PlanAdmissionOperation]

File: javascript/src/credits/types/catalog.ts
import type { Decimal } from "decimal.js";

export interface CatalogRevision {
  id: string;
  config: Record<string, unknown>;
  version: number;
}

export interface CatalogRevisionSummary {
  id: string;
  version: number;
  label: string | null;
  active: boolean;
  createdAt: string;
}

export interface PlanAllowancePolicy {
  amount: Decimal;
  priority: number;
  resetUnit: "second" | "minute" | "hour" | "day" | "week" | "month" | "year";
  resetCount: number;
  resetAnchor: "calendar" | "plan_assignment" | "rolling";
  resetTimezone: string;
}

export interface PlanCreditPolicy {
  type: "prepaid" | "credit_line";
  creditLimit: Decimal | null;
}

export interface PlanAdmissionPolicy {
  maxInFlight: number | null;
  operations: Record<string, { maxInFlight: number | null }>;
}

export interface AllowanceResult {
  planId: string;
  allowanceRemaining: Decimal;
  periodStart: string;
  periodEnd: string;
}

export interface GetUserPlanResult {
  userId: string;
  planId: string | null;
  planKey: string | null;
  planLabel: string | null;
  allowance: PlanAllowancePolicy | null;
  entitlements: Record<string, { value: unknown }>;
  rateCard: string | null;
  creditPolicy: PlanCreditPolicy | null;
  admission: PlanAdmissionPolicy | null;
  allowedOperations: string[];
  planAssignedAt: Date | null;
  planAssignmentEndsAt: Date | null;
  assignmentSourceType: "manual" | "subscription" | "migration" | "system" | null;
  assignmentSourceId: string | null;
  catalogRevisionPinned: boolean;
  catalogVersion: number | null;
}

export interface SetUserPlanResult {
  userId: string;
  planId: string;
  planKey: string;
  planAssignedAt: Date;
  assignmentState: "applied" | "scheduled";
}

export interface UnsetUserPlanResult {
  userId: string;
}

export interface PlanMigrationStartResult {
  migrationId: string;
}

export interface PlanMigrationBatchResult {
  migrated: number;
  done: boolean;
  nextCursor: string | null;
}
